### ISSUE TITLE
ToDo: followsups / extras / ideas [cont.... again]

### ISSUE THREAD

**Issue by Thorin-Oakenpants** [ghacksuserjs/TorZillaPrint#54]
previous: #28, #46

---

### :small_red_triangle_down: ToDo

:small_blue_diamond: screen
* [ ] screen: update/add border-snapping subPixel devicePixelRatio PoC to use domrect
* [ ] screen: improve android keyboard PoC to use setInterval to detect a static change
* screen: android 73+: see [1514429](https://bugzilla.mozilla.org/1514429):
   * [ ]  drop toolbar height test?
   * [ ] anything else affected?
* [ ] screen: android: investigate a history of changes as soon as possible
   - e.g global.js is where I grab initial viewport and window sizes
* [ ] screen: android: exploit `font inflation`
* [ ] screen: android: exploit position (top/bottom) of toolbar in Fenix [yes, I can detect this]

:small_blue_diamond: ua
* ua: worker/iframe nav checks
   * [ ] add nested worker & worker from blob & all iframe tests
   * [ ] make all tests handle failures and  promise/resolve into a single array: output unique leaks
   * [ ] combine/split duplicate/diff hashes per type: workers / iframes
* [ ] ua: finish UA BS detection
   - this is implausible results based on feature detection and typos/spaces etc and accounts for RFP

:small_blue_diamond: math
* [ ] math: build a matrix (ECMAScript 1st vs 6th) of all results
   - retest, create/test more Linux VM distros
   - 99% sure this is all correct, but may find new hashes

:small_blue_diamond: language
* [ ] language: investigate `getCanonicalLocales`

:small_blue_diamond: storage
* [ ] storage: finish other tests

:small_blue_diamond: devices
* [ ] devices: finish pointer tests
* [ ] devices: android: touch tests: `radiusX/Y`, `screenX/Y`, `clientX/Y`
* [ ] devices: enumerate gamepads
* [ ] devices: enumerate vrdisplays
* [ ] use precision timing bypasses to estimate cores

:small_blue_diamond: domrect
* [ ] domrect: investigate other tests: see CB extension options

:small_blue_diamond: canvas
* [ ] canvas: add offscreenCanvas
* [ ] canvas: add iframe tests
* [ ] canvas: add `noise` detection: i.e detect stealth as compared to random on each pass
   - isPoint left to do
* [ ] canvas: add `noise` fingerprint... maybe
* [ ] canvas: get size? get number of colors? see #60 
* [ ] canvas: decide if stability is important see #60 
   - the canvas content is causing changes between some FF releases whereas CB's doesn't

:small_blue_diamond: webgl
* [ ] webgl: readPixels: expand/test on error entropy
   - e.g. in VM's which usually block the graphics driver and the error messages can differ
   - make sure error entropy is handled for TB slider and VMs
* [ ] webgl: build it: image1+image2, parameters, colors, size, iframe tests

:small_blue_diamond: audio
* [ ] audio: investigate other tests: see CB

:small_blue_diamond: fonts
* [ ] fonts: unicode gylphs: textMetrics: fixup width randomization dectection
* [ ] fonts: unicode gylphs: textMetrics: exploit double char non-matching for extra entropy
* [ ] fonts: fontfallback: investigate further
   - when document fonts are blocked, the first run misses fonts it picks up on subsequent tests: currently mitigated by doing a 2-font "priming" run: but this is not an elegant solution
* [ ] fonts: build lists for mac, linux, android - see #43
* [ ] fonts: graphite
* [ ] fonts: default font-family name: serif + san-serif
* [ ] fonts: font versions: Times New Roman, Arial [thru unicode changes per version]

:small_blue_diamond: media
* [ ] media: improve audio/video lists
   - reduce list for perf (we only care about FF)
   - add items for entropy
* [ ] media: image hash: encoding/decoding FP?
* [ ] media: mediacapabilities FP

:small_blue_diamond: misc
* [ ] misc: investigate: window properties objects hash

:small_blue_diamond: other
* [ ] other: WebRTC
* [ ] other: isAdBlocking [also: hash of results]
* [ ] other: santitzing / zombie storage - [prototype](https://ghacksuserjs.github.io/TorZillaPrint/sanitizing.html)
* [ ] other: widgets/form controls
* [ ] other: RFP timing tests - [temp demo / ideas](https://thorin-oakenpants.github.io/testing/timing.html)
* [ ] other: elements
* [ ] other: AOM (accessability object model)
* [ ] other: SVG
   - SVGTextContentElement.prototype.getComputedTextLength
   - SVGPathElement.prototype.getTotalLength
   - SVGGeometryElement.getPointAtLength()

:small_blue_diamond: randomizing detection
* [ ] rando: audio / ~~domrect~~ / measureText / ~~canvas (see canvas noise)~~ etc
* [ ] rando: ^^ might be useful for css line-height / unicode gylphs 

---

### Done :small_red_triangle_down:

Mar-13-2020
- [x] ua: added `non-standard` notation to `error messages` and `error2` if user flips `javascript.options.property_error_message_fix` from default
   - this is hash based and handles the case regardless of the default value
- These are pretty much already covered
   - language: ~~`Intl.NumberFormat`: add more types e.g. fraction?~~
   - language: ~~`Intl.NumberFormat formatToParts`: add more types e.g. currency?~~
   - language: ~~`Intl.NumberFormat formatToParts BigInt`: FF70+~~
* [x] ua: element scrollbar width: highlight when viewport !== element
   - added notation at 100% zoom level
   - outside 100% zoom there can be legitimate differences

Mar-18-2020
- [x] css: expanded system colors to include CSS4 system colors: [FF76+ 1590894](https://bugzilla.mozilla.org/show_bug.cgi?id=1590894)
- [x] css: expanded `-moz-` colors: from 36 to 61 items
   - also removed one duplicate and one item deprecated in FF26
- [x] misc: added a **basic** RFP timing check
   - it uses `performance.now()` over 130 ms (1 result every 13 ms)
   - all 10 results must = `0` when `% 100`
   - its entirely feasible that the `100ms` RFP default (which is behind prefs) could be altered to over 130, but I will build in more complex tests in the future - see above

Mar-22-2020
- [x] media: video hash: canPlayType, isTypeSupported (MediaSource + MediaRecorder)
- [x] media: audio hash: canPlayType, isTypeSupported (MediaSource + MediaRecorder)

Mar-24
- [x] fonts: split details into two sections - autorun vs manual 
- [x] fonts: expand textMetrics
   - overall textMetric hash contains entropy for canvas 2d being blocked vs the method not available
   - all the new methods (FF74+, i.e everything except width) are behind prefs, and at this stage I cannot tell if it is the pref or an extension like CB causing a failure, so I can only fall back to the default pref value - which is also baked into the overall hash: at least for `actualBounding`
  - for the remaining seven methods I will add this when they get turned on

Mar-25
- [x] fonts: textMetrics: detection of randomizing in `width`
   - edit: this needs more work

Mar-30
- [x] ua: fixed window.open navigator check to use the right window
- [x] ua: added worker navigator check
- [x] ua: added shared worker navigator check
- [x] storage: service worker check is now unregistered

Apr-01
- [x] ua: added service worker navigator check

Apr-05
- [x] css: split CSS4 system colors out with details
   - I can't provide an RFP green check for (older) system colors with the CSS4 ones, since the new ones leak entropy across platforms, setups or whatever (it's not looking pretty)
* [x] language: added `Intl.RelativeTimeFormat` + `numberingSystem`
* [x] language: added `Intl.DateTimeFormat` + `dayPeriod`: [1569103](https://bugzilla.mozilla.org/1569103)

Apr-06
- [x] language: added `Intl.DateTimeFormat` format to parts + `relatedYear` + `yearName`

Apr-15
- [x] fonts: textMetrics
   - added supported vs non-supported vs default values
   - added supported vs blocked
   - notation added for when not standard pref value
   - ^^ all the above incorporated into the overall hash
  - details still output when canvas is blocked (i.e if it's supported or not and if not standard config)
- [x] ua: version 77+ detection

Apr-16
- [x] usability: anchors
   - clicking down nav now keeps you on the right hand side of the page
   - e.g. if you've pinched to zoom on android and are jumping sections
  - note: the up nav and section title clicks keep you ~~right~~ left aligned: no change

May-09
- [x] languages: differentiate between default geo settings for TB vs FF
- [x] misc: added windows.content
- [x] misc: added PerformanceNavigationTiming [1511941](https://bugzilla.mozilla.org/show_bug.cgi?id=1511941) with RFP notation for FF78+
   * this is a very good fast test for detecting RFP (with isFF)

May 10
- [x] languages: detect timezone/languages spoofs via worker

May 11
- [x] ua: when testing window.open, output leaks if not already exposed
   - I already plan to make all iframe/worker tests into promises and resolve leaks in a single function, and I might do same as CB and include window.open - i.e don't auto-run, just add a run button for entire sub-section
- [x] ua: show/hide vendor/vendorSub
   * this is already in my BS detection function (which detects BS based on feature detection, typos, incorrect spacing, etc and which allows for RFP), but added in the meantime. My BS detection function is not ready yet :)
- [x] ua: detect when a navigator property is blocked
   * a zero-length string is returned as "undefined" (e.g. vendorSub)
   * `undefined` (non-string) is returned as "blocked": I only do this for FF (as other browsers don't always have the same navigator objects)
* [x] storage: appcachetest: decide what to do: the storage is deprecated in [FF71+](https://bugzilla.mozilla.org/show_bug.cgi?id=1237782) but not the API, and TB blocks it anyway.
   - removed all references to it: it's just not worth testing: entropy from how FF handles it is already covered by version detection and other extras like branding / specific Nightly detection, etc.

May-13
- [x] from Mar-13: ua: `javascript.options.property_error_message_fix` default state/notation
  - now caters for the change in 77+ developer: see [1636195](https://bugzilla.mozilla.org/show_bug.cgi?id=1636195)
   - i.e I distinguish between dev vs beta and apply a different logic in FF77+
   - note: I can't do this on android (no branding to leak the release channel): so I ignore it

May-15
- [x] canvas: does a second test, and RFP notation in 78+ accounts for randomizing on some metrics

May 16
- [x] domrect: no longer uses an iframe
   - instead it uses a hidden fixed overlay
   - no waiting, one less file to fetch, no cleaning up, not trying to handle blocked iframes
   - if I need to use the iframe version: that's updated to match except the hidden part: tested and hashes match
   - perf massively improved
- [x] domrect: error handling to detect blocking of each method
- [x] canvas: misc
   - improved output logic for notation
   - canvas has an overall hash: if random it records that rather than the hashes. This way a repeat test of RFP (e.g say an iframe) will produce the same hash
   - the hash and notation is different when I can detect an `RFP random` vs an `extension random` (if the extension's 2nd result is not the same: i.e not in stealth mode: e.g. CB's use canvas cache)

May 17
- [x] canvas
   - even MOAR and better logic and MOAR checks
      - e.g randomized check for `toDataURL` vs `toBlob`
   - did away with adding a random notation, and instead output the two hashes on one line
   - fixed issues with non-FF (maybe they can also happen in FF?) where sometimes toBlob gets an error on the first test but not the second: when this happens I use the error only: since the test requires two valid results to be valid (e.g. detecting randomness)
   - tested like F
       - pre74 (mozGetAsFile), 77 (old RFP logic), 78 (new RFP logic)
       - in 78 with RFP randomizing on and off
       - in 78 with RFP on but a site exception with CB kicking in (this is not the same as RFP off: as I use this distinction to show a bad RFP random if CB is not cached)
       - CB with various settings (and there's a heap of them)
       - ^^ all the above in various combos

May 23
- [x] ua: version test73 replaced
   - test 73 was the only one that relied on css. If css was blocked (in this case it's inline), then the test would fail
   - I also built a version test page to keep all my version PoCs for posterity, used on TZP or not - it's [here](https://thorin-oakenpants.github.io/testing/versions.html). And I created sections and timed them for perf just as they would be used on TZP
- [x] misc: new RFP timing checks
   - `[getEntries*] performance.mark` and `[loadEventEnd] performance.timing`
   - getEntries seems to be impervious to timing pref changes such as
       - > dom.enable_performance_navigation_timing, dom.enable_performance_observer, dom.enable_performance, dom.enable_resource_timing

May 24
- [x] languages: added `Intl.ListFormat`

Jun 8
- [x] languages: added `Intl.ListFormat` locale

Jun 10
- [x] languages: added dateStyle, timeStyle FF79+

May-Jun
- [x] ua: version 78, 79+ added
- [x] ua: version tests now cascade if needed, and every release has a test
   - the error hash is used to know where to start in the version checks (for perf): but scripts can mess with the error hash, or indeed, the error hash could change. When this happens, the version checking cascades down, across functions until it gets a successful result
- [x] ua: version 78 has multiple checks (to help counter script blocking interference). 78 is important for implementing the upcoming RFP UA BS Pinocchio Meter

Jun 16-21
- [x] languages: added `getTime` for timezones
- [x] languages: everything except `geo` is ~~not~~ now in the worker leak test
- [x] domrect: add randomization detection (because I'm a clever bastard)
- [x] languages: added `Date.parse` for timezones

Jun 24-25
- [x] languages: added `Geolocation` in `window` for geo: this doesn't add any entropy overall since it's version dependant
- [x] languages: worker test: fixed errors (missing variables/functions) and added in more script blocking detection (for item 26)
- [x] css: added css [system fonts](https://dxr.mozilla.org/mozilla-central/source/servo/components/style/properties/longhands/font.mako.rs#320)

Jun 27
- These don't add anything
   * ~~language: investigate `Intl.Locale`~~: FF75+: [1613713](https://bugzilla.mozilla.org/1613713), [MDN](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Intl/Locale)
   * ~~language: investigate `Intl.DisplayNames`~~: [1557727](https://bugzilla.mozilla.org/1557727) [MDN](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Intl/DisplayNames)

...

**Comment by Thorin-Oakenpants** [comment]
@sereneblue FYI
@kkapsner tech question coming up

So I added a worker test to timezone and languages. Here I am on an en-US build in my own timezone (no RFP), and using Chameleon (current AMO version: which I understand will have some teething issues after the rewrite) I am spoofing language as German and timezone as UTC. These are the only two things I'm spoofing (but sereneblue, note that user agent also leaks via workers)

Here's a pretty picture (blocked out my timezone, sorry guys)
![tz-lang](https://user-images.githubusercontent.com/16656956/81517408-6889eb00-932a-11ea-96b6-59eb8f11a3d9.png)

Note this doesn't include anything to do with the `date/time format` section, but that has lots of leaks without workers.

- In this setup (language + timezone spoof) `[Int.PluralRules] locale` was not spoofed
- if I turn off the timezone spoof `[resolvedOptions] locale` is **also** not spoofed
- so not sure why timezone spoofing option affects a locale

@kkapsner ... here we go
So I did this via a worker. I tried to use a shared worker (because that seemed to leak more than a worker in the UA tests? I forget now), but it just wouldn't work


workershared_lang.js
```
var ports = []
onconnect = function(e) {
	console.debug("testing: onconnect")
	let port = e.ports[0]
	ports.push(port)
	port.start()
	port.onmessage = function(e) {
		console.debug("testing: onmessage")
		let res = [],
			d1 = new Date("January 30, 2019 13:00:00"),
			d2 = new Date("July 30, 2018 13:00:00")
		let tz1 = d1.getTimezoneOffset()+ ' | ' + d2.getTimezoneOffset()
		let tz2 = Intl.DateTimeFormat().resolvedOptions().timeZone
		res.push(tz1)
		res.push(tz2)
		res.push(navigator.languages)
		res.push(navigator.language)
		res.push(navigator.languages[0])
		res.push(new Intl.PluralRules().resolvedOptions().locale)
		res.push(Intl.DateTimeFormat().resolvedOptions().locale)
		port.postMessage(res)
	}
}
```

^^ see that, looks legit. I could get the onconnect debug, but not the onmessage

how I call it
```
		try {
			let sharedlang = new SharedWorker("js/workershared_lang.js")
			sharedlang.port.addEventListener("message", function(e) {
				// do stuff
			}, false)
			sharedlang.port.start()
			sharedlang.postMessage("hi")
		} catch(e) {}
```

this didn't work: but it works for the ua ones. So WT heck man? :) I'm clearly missing something

**Comment by kkapsner** [comment]
> because that seemed to leak more than a worker in the UA tests? I forget now

The normal Worker is weird in regards of the userAgent.

I do not see that you close your port to the SharedWorker. Maybe you then have more than one port. So check in the onconnect function what is all in `e.ports`.
Also check if something lands in the `catch` block.

**Comment by Thorin-Oakenpants** [comment]
Ooh . I have at least 2 other shared worker tests (one in cookies, one in ua) ... so I better learn how to close ports then

**Comment by Thorin-Oakenpants** [comment]
> The normal Worker is weird in regards of the userAgent

Not sure what weird means - all I know is it detected 5 spoofs, not 6 compared to shared worker. Anyway, I'm going to stick with worker in here (it seems everything leaks that I need it to) because a web worker can still be created and used when cookies are blocked (site permissions) whereas a shared worker can't

**Comment by Thorin-Oakenpants** [comment]
@kkapsner 

You could tidy the manual input into nav spoofing values by automatically
- removing leading and trailing spaces
- change multi-space into single-space
- if blank after trimming then re-insert `{original value}`
- that's for custom values obviously
- ^^ my BS detection function picks up on those: unless there are actually nav properties that have those: but certainly not in FF

here's an example
- platform with a trailing space
- appVersion with an extra space in the middle
- appCodeName with a leading space

this is the live version showing the diff from the worker tests
 - it's hard to tell from the platform one (trailing space). I used to wrap the output in quote marks but it looked silly/misleading
![CBnav4](https://user-images.githubusercontent.com/16656956/81628658-c54fd900-93f0-11ea-9632-18ddedab3f7c.png)

this is my local file:// version (so no workers), but with my awesome BS detector (patent pending!! lulz) - it also handles case (mozilla vs Mozilla etc), and lots more, but I think that's out of your spec
![CBnav5](https://user-images.githubusercontent.com/16656956/81628886-6b034800-93f1-11ea-87ab-dbe673a92c2a.png)

---

So I now detect blocked vs undefined etc in ua nav objects (see Mar-11 changes in first post)

Using CB set to block everything (the global setting in general block mode)

Here's the property options
![CBnav1](https://user-images.githubusercontent.com/16656956/81627932-cbdd5100-93ee-11ea-8b36-58ff3f4b25bc.png)

Here's my spoofing setup
![CBnav2](https://user-images.githubusercontent.com/16656956/81627966-e1527b00-93ee-11ea-9f9e-4c6feae09566.png)

Here are the results
![CBnav3](https://user-images.githubusercontent.com/16656956/81627972-e8798900-93ee-11ea-934d-068b7ce97886.png)

What I find interesting is that you can only block or fake: not block some, fake the others :) That and the block mode is global per API, not per API

**Comment by Thorin-Oakenpants** [comment]
^^ anyway: blocking is just silly (as long as I can code it), so mix and matching fake + block within an API is a moot point.

But faking vs blocking **per** API is possibly a legit improvement?

**Comment by sereneblue** [comment]
@Thorin-Oakenpants 

Thanks for the heads up. It seems there is a bug with timezone spoofing that I'll need to fix. I've never actually used Workers before although I have heard of them; didn't realize you could access the navigator properties there too. I'll see if I can address in a future update.

**Comment by Thorin-Oakenpants** [comment]
@sereneblue I'm not an extension dev, but the way I understand it, is that the APIs to tackle workers (web workers, shared workers, service workers, nested workers ... but especially workers from blobs) is limited if not non-existent .. so whatever a worker is allowed to do in JS, will leak.

**Comment by Thorin-Oakenpants** [comment]
@sereneblue I think you also leak on window.open

**Comment by kkapsner** [comment]
> Not sure what weird means - all I know is it detected 5 spoofs, not 6 compared to shared worker.

In a nested worker it should also detect 6.

>     * I used to wrap the output in quote marks but it looked silly/misleading

You could replace the space by a point (https://en.wikipedia.org/wiki/Interpunct)

> You could tidy the manual input into nav spoofing values by automatically

hm... could do that - not sure if it's worth the effort. I think it would be better to not change these automatically but to add a sanitation rule for these.

> But faking vs blocking **per** API is possibly a legit improvement?

Could be. Do you have a specific use case (but not testing TZP... ;))?

@sereneblue I did not find a feasible way to interact with workers so I opened https://bugzilla.mozilla.org/show_bug.cgi?id=1636977 - let me know if you found a way.

> @kkapsner can probably correct me: but I believe it's currently limited to the navigator API ?

Workers have a limited set of APIs: https://developer.mozilla.org/en-US/docs/Web/API/Web_Workers_API/Functions_and_classes_available_to_workers
At the moment only the navigator API is relevant for CB. But when the offscreen API gets enabled/implemented (it's a hot mess at the moment) canvas fingerprinting will be possible in workers.

**Comment by Thorin-Oakenpants** [comment]
> Workers have a limited set of APIs ...

So I found a way for workers to detect isFF .. I'll just pass it along in the message I send :)

**Comment by Thorin-Oakenpants** [comment]
> In a nested worker it should also detect 6

Cool. I still need to wrap up all the iframe/worker tests and add the ones missing, into a single promisified thingy (I need your help!!) - I know shared and service do six.

I said workers still worked with cookies blocked: but in the past that hasn't been the case. I just retested in Nightly, and they do work. Weird.

> Workers have a limited set of APIs ...

So I found a way for workers to "detect" `isFF` (running the function to check for Firefox doesn't work in workers) -  I only want to return "blocked" for Firefox because not all browsers return values for all nav objects - and I want all my nav checking to return the same values for when I compare them for leaks - I'll just pass it along in the message I send :)

> Do you have a specific use case

Nope... just keeping you busy and on your toes

> but to add a sanitation rule for these

Isn't that what I proposed? I'm not an ext dev, but I would assume you just pass the string from each field thru a function before saving. Or I guess that word means something different.

I actually quite like it the way it is: so I can easily test my AWESOMEST BEST-EVER MOST-BIGLY PATENT-PENDING BS & PINNOCHIO DETECTOR ... I want to point out to people how useless spoofing the UA is

> middot

I'll see what it looks like

**Comment by Thorin-Oakenpants** [comment]
seems OK

Not sure if it's worth it just for trailing spaces, which probably almost no-one would have. It would be awesome if the spoof and real value were the same except the trailing space, and they're like "woah, how did he know that I was spoofing that", but then again, they're just as likely to think "this test is faulty". A double space can also be missed.

Maybe the real solution is to monospace it and replace all spaces with a middot.

What do you think? I'm thinking leave it as it (no middots)?

![middot](https://user-images.githubusercontent.com/16656956/81696253-2e683880-9453-11ea-8a42-54c6a1f838dc.png)

**Comment by Thorin-Oakenpants** [comment]
@kkapsner .. so I guess this is right, right?
![cleaning-up](https://user-images.githubusercontent.com/16656956/81711390-27492680-9463-11ea-8b26-7b6ee524fd32.png)

**Comment by kkapsner** [comment]
> Isn't that what I proposed? I'm not an ext dev, but I would assume you just pass the string from each field thru a function before saving. Or I guess that word means something different.

No - that word means also that. But in CB I have a separate page that checks the settings for bad configuration. This is called "Setting sanitation". It simply outputs warnings for bad configurations. 

> is it really worth it?

not sure... a monospace font would also help to see the spaces. Maybe in combination with a light gray border.

> so I guess this is right, right?

Look fine - does it work?

**Comment by Thorin-Oakenpants** [comment]
> It simply outputs warnings for bad configurations.

Ah. I can see the point of that: conflicts etc across multiple settings (and the user has to actively go look at that sanitize section, right? and click a button? soz for not looking myself). So for something as simple as custom string cleaning for one little section of the UI, I think it's better to fix it at the source: i.e before saving

of course, your choice :) I'd like it left "as is" for my testing, but then I'm a selfish bastard sometimes

> monospace .. background

not feeling it, at all :) monospace takes more space, and that causes word wrap. Not keen on disrupting the color (left/right border). I don't actually check for leading/trailing/multi-spaces - they just get auto picked up on with my award-winning pinnochio algorithm - otherwise I could append a pale grey `[trailing space]` or something: I could specifically check for those three things just to add the decoration. But then that too just eats space.

I'll keep thinking about it: your suggestions are all good ideas.

> Look fine - does it work?

Did to the storage tests (worker, shared-w), and the timezone-language worker as well - and there are no errors. Service workers I already close the broadcast and unregister.

**Comment by Thorin-Oakenpants** [comment]
@kkapsner FYI: [1621433](https://bugzilla.mozilla.org/1621433) landed an hour ago

**Comment by Thorin-Oakenpants** [comment]
@kkapsner hah,  the [maximum depth for content iframes](https://searchfox.org/mozilla-central/rev/5a10be606f2d76ef22f1f44565749490de991d35/dom/base/nsFrameLoader.cpp#140-147) is 10

Does your extension handle a nested-nested-nested-nested-nested-....-nested iframe?

**Comment by kkapsner** [comment]
> and the user has to actively go look at that sanitize section, right? and click a button?

Yes

> I think it's better to fix it at the source

I'm not sure. There might be use cases (in the future) where these weird inputs make sense.

> Does your extension handle a nested-nested-nested-nested-nested-....-nested iframe?

Did not test it exhaustively but it should.

**Comment by Thorin-Oakenpants** [comment]
For your info, for anyone playing at home... took me almost 3 hours to find a new test for FF73. The old one relied on css (so if css was blocked, the test would fail).

In doing this, I decided to stick all my version PoCs in one place: a lot aren't used since I moved to leveraging the error hashes and speed up the process for version detection.

https://thorin-oakenpants.github.io/testing/versions.html <-- have at it

It doubles as a check (e.g. if something passes at a higher value than it should: example: Tor Browser ESR60 based passes the 73b test), and also as a perf check for how it would happen on TZP


---

Actually reading just above that link it says
```js
// Bug 136580: Limit to the number of nested content frames that can have the
//             same URL. This is to stop content that is recursively loading
//             itself.  Note that "#foo" on the end of URL doesn't affect
//             whether it's considered identical, but "?foo" or ";foo" are
//             considered and compared.
// Limit this to 2, like chromium does.
#define MAX_SAME_URL_CONTENT_FRAMES 2
```

Not sure how that affects your testing

**Comment by kkapsner** [comment]
> Not sure how that would affects your multiple nested testing

Should not affect it - but thanks for the info.

**Comment by Thorin-Oakenpants** [comment]
@sereneblue You might want to update language spoofing to cover Intl.ListFormat resolvedOptions locale (I'm spoofing in Chameleon as de)

![eek](https://user-images.githubusercontent.com/16656956/84080033-112f7700-a9cb-11ea-932c-fac0fdb24144.png)

**Comment by sereneblue** [comment]
@Thorin-Oakenpants Thanks for the heads up. This will be fixed in the next update before FF78 rolls out.

**Comment by Thorin-Oakenpants** [comment]
@sereneblue I just updated Chameleon (20.15) .. there seem to be some regressions

![regression](https://user-images.githubusercontent.com/16656956/85095720-bf17fe00-b1e1-11ea-9291-76a4f06096d7.png)

**Comment by sereneblue** [comment]
@Thorin-Oakenpants Thanks, this will be fixed in v0.20.16.

**Comment by Thorin-Oakenpants** [comment]
@sereneblue tested: looks good except `Intl.ListFormat | script blocked [c]` is still there

also, I assume you read #63 - e.g. getTime leaks, and dayPeriod code is a bit whack (gives double ups of various periods) but also dayPeriod is a specific date/time and shouldn't change: not sure how you can handle that: i.e differentiate between a specific date/time vs e.g. date.now() ?

**Comment by Thorin-Oakenpants** [comment]
FYI: @sereneblue I just added `Date.parse` to timezones, and they leak as well. I don't know if that's an easy fix or not: I think I read @kkapsner say somewhere that you'd have to build your own date parser?

**Comment by kkapsner** [comment]
> I think I read @kkapsner say somewhere that ~you'd~ he'd have to build ~your~ his own date parser if he was going to do it?

Yes. E.g. if you run `new Date("2020-01-01T12:00:00.000")` it will return noon in your time zone. So if you fake the time zone you have to parse this input (according to spec and according to the browser quirks that may be present). So a way to get the timezone offset with `.parse` is:
```
const nowISO = new Date().toISOString();
console.log((Date.parse(nowISO.replace("Z", "")) - Date.parse(nowISO))/60000);
```

**Comment by sereneblue** [comment]
@Thorin-Oakenpants Are you testing `Intl.ListFormat` in FF77? It's a new feature in FF78. Chameleon checks if it exists before spoofing it. I did see that other thread. I'm working on a potentially simpler implementation of timezone spoofing that should resolve the leak with `getTIme` and `Date.parse`. It's possible to write a wrapper for Firefox's`Date.parse` method; really depends on how deep you want to go in the rabbit hole.

**Comment by Thorin-Oakenpants** [comment]
They all try catch, so it's feature detected. `Intl.ListFormat` the one that returns `a, b, and c | a, b, or c | a, b, c |  etc` is item 36

```js
	} else if (item == 36) {
		// ListFormat: 1589095: 78+
		let tmp36 = "", res36 = []
		let	styles = ['long','short','narrow'],
			types = ['conjunction', 'disjunction','unit']
		styles.forEach(function(s){
			types.forEach(function(t){
				res36.push(new Intl.ListFormat(undefined,{style: s, type: t}).format(["a","b","c"]))
			})
		})
		if (res36.length > 0) {tmp36 = res36.join(" | ")}
		return tmp36
	}
```

and the catch contains for Firefox only
```js
	if (item == 5 || item == 36) {
		if (e.message == "Intl.ListFormat is not a constructor" && isVer < 78) {msg = zNS}
	}
	// script blocking

		// script blocking
		if (msg == "") {
			err.push(item +" [unexpected]: "+ e.name + " : " + e.message)
			if (e.name == "ReferenceError") {return zB1
			} else if (e.name == "TypeError") {return zB2
			} else {return zB3}
		}
```

`zNS` = "not supported`. So it's feature detected, but only allows the correct FF error message in 77 and lower, otherwise it's being affected by a script.

`zB3` = `script blocked c`, so it's causing an error, and you should see it in the console (I output `err.join("\n")`

There's also multiple checks to make sure you're on 78 (as a minimum, my isVer works down). What does your version say (look at the UA section)

**Comment by Thorin-Oakenpants** [comment]
running a test in Nightly 79, I get
```
language/datetime errors
36 [unexpected]: RangeError : invalid language tag: "[object Object]"
```

I'd have to install Chameleon in a 77 and a 78 to test further

**Comment by kkapsner** [comment]
> It's possible to write a wrapper for Firefox's`Date.parse` method; really depends on how deep you want to go in the rabbit hole.

You also need to replace the Date constructor as it can also parse strings.

**Comment by sereneblue** [comment]
> > It's possible to write a wrapper for Firefox's`Date.parse` method; really depends on how deep you want to go in the rabbit hole.
> 
> You also need to replace the Date constructor as it can also parse strings.

Indeed, it can quite complicated. 

> They all try catch, so it's feature detected. `Intl.ListFormat` the one that returns `a, b, and c | a, b, or c | a, b, c | etc` is item 36
> 
> ```js
> 	} else if (item == 36) {
> 		// ListFormat: 1589095: 78+
> 		let tmp36 = "", res36 = []
> 		let	styles = ['long','short','narrow'],
> 			types = ['conjunction', 'disjunction','unit']
> 		styles.forEach(function(s){
> 			types.forEach(function(t){
> 				res36.push(new Intl.ListFormat(undefined,{style: s, type: t}).format(["a","b","c"]))
> 			})
> 		})
> 		if (res36.length > 0) {tmp36 = res36.join(" | ")}
> 		return tmp36
> 	}
> ```
> 
> and the catch contains for Firefox only
> 
> ```js
> 	if (item == 5 || item == 36) {
> 		if (e.message == "Intl.ListFormat is not a constructor" && isVer < 78) {msg = zNS}
> 	}
> 	// script blocking
> 
> 		// script blocking
> 		if (msg == "") {
> 			err.push(item +" [unexpected]: "+ e.name + " : " + e.message)
> 			if (e.name == "ReferenceError") {return zB1
> 			} else if (e.name == "TypeError") {return zB2
> 			} else {return zB3}
> 		}
> ```
> 
> `zNS` = "not supported`. So it's feature detected, but only allows the correct FF error message in 77 and lower, otherwise it's being affected by a script.
> 
> `zB3` = `script blocked c`, so it's causing an error, and you should see it in the console (I output `err.join("\n")`
> 
> There's also multiple checks to make sure you're on 78 (as a minimum, my isVer works down). What does your version say (look at the UA section)

I see, I've fixed this issue. It should be resolved in the next Chameleon update.

> > It's possible to write a wrapper for Firefox's`Date.parse` method; really depends on how deep you want to go in the rabbit hole.
> 
> You also need to replace the Date constructor as it can also parse strings.

Chameleon does replace the Date contstructor but it doesn't handle the timezone strings well. There's still some work that needs to be done to better handle spoofing.

**Comment by kkapsner** [comment]
> Chameleon does replace the Date constructor

Well, this is very easy to detect: 
![image](https://user-images.githubusercontent.com/1105157/85918555-2d089900-b864-11ea-9c28-11677cffebf8.png)

**Comment by sereneblue** [comment]
> > Chameleon does replace the Date constructor
> 
> Well, this is very easy to detect:
> ![image](https://user-images.githubusercontent.com/1105157/85918555-2d089900-b864-11ea-9c28-11677cffebf8.png)
> 
> EDIT: also `Date.length` (it should be 7) and `Date.name` (it should be "Date") are incorrect
> EDIT 2: `Date.arguments` and `Date.caller` should throw an error (i.e. you have to make the function strict).
> EDIT 3: not a problem in Firefox but Chrome: https://developer.mozilla.org/en-US/docs/Web/API/File/lastModifiedDate returns a Date object from which you can extract the constructor (or just check with `instanceof` if the Date constructor has been changed). But there might be some other JS API that returns a Date object. You have to check/alter all of them.

Thanks. These issues should be addressed in the recent update. You've brought up in interesting scenario with your third edit. I'll investigate a bit further and will try to fill the gaps where I can.

**Comment by kkapsner** [comment]
Still detectable: `Date.toString === Function.prototype.toString`
![image](https://user-images.githubusercontent.com/1105157/86773195-9b640d00-c055-11ea-8b30-b7af11683d1c.png)

**Comment by Thorin-Oakenpants** [comment]
@ anyone .. I'm having a little trouble with [this](https://bugzilla.mozilla.org/show_bug.cgi?id=1651732) as a potential v80 check

```js
	// 80: 1651732
	try {
		let obj80 = {
			Symbol.toPrimitive {
				return(Symbol())
			}
		}
		let proxy = (new Proxy({},{
			get(obj80, property, proxy) {
				return(property)
			}
		}))
		try {
			console.log(11, proxy[object])
			// no error: you're on 80+
		} catch (err) {
			// TypeError: can't convert symbol to string
			// error: you're on 79 or lower
		}
	} catch(e) {
		console.debug("80new", e.name, e.message)
	}
```

I get a `Uncaught SyntaxError: missing : after property id` for the line `Symbol.toPrimitive {` . I'm also assuming that this could be simplified

[MDN: Symbol.toPrimitive](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Symbol/toPrimitive)

**Comment by abrahamjuliot** [comment]
Looks like `Symbol.toPrimitive` should be within a method dynamic property key.

```javascript
obj80 = {
  [Symbol.toPrimitive]() {
    return (Symbol())
  }
}
// or es5
obj = {}
obj[Symbol.toPrimitive] = () => Symbol()
```

**Comment by Thorin-Oakenpants** [comment]
Hmmm, I thought I had tried that, and it doesn't catch err. But I think I can use it: 80+ returns 11, 79 or lower returns null. I just to confirm and check the perf

**Comment by abrahamjuliot** [comment]
This error seems to be triggered on the 11th iteration of a loop. Here are 2 examples checking if typeof is 'symbol':

```javascript
const obj= {
  [Symbol.toPrimitive]() {
    return Symbol()
  }
}
const proxy = new Proxy({},{
  get(obj, prop, proxy) {
    return prop
  }
})

const len = 200 // any length greater than 11
const test = []
for (let i = 0; i < len; i++) {
  try {
     if (typeof proxy[obj] == 'symbol') {
       test.push(true)
     }
  }
  catch (error) { }
}

console.log(test.length) // '79' is <= 11
```

```javascript
const obj = {
  [Symbol.toPrimitive]: () => Symbol()
}
const proxy = new Proxy({},{
  get: (obj, prop, proxy) => prop
})

const len = 200 // any length greater than 11
const test = []
;[...Array(len)].forEach((val, i) => {
  try {
     if (typeof proxy[obj] == 'symbol') {
       test[i] = true
     }
  }
  catch (error) { }
})

console.log(test.length == len) // '80' is true
```

**Comment by Thorin-Oakenpants** [comment]
👍 Ahh .. it's only after tries hits double digits : bugzilla OP said when `> 9`:  i.e `0-9` = ok, `10+` = not okay

I'll see how this goes
